### PR TITLE
[RFC] Add TemplateType

### LINKED ISSUE #5863: [RFC] TemplateType Integration
## Feature Request
I  would like to provide a new FormType called `TemplateType`. This allows you to add your custom template into the SonataAdmin configureFormFields without overriding any sonata template.

Example

Admin:
```php
protected function configureFormFields(FormMapper $formMapper): void
{
      $formMapper
          ->with('default')
               ->add('example', TemplateType::class, [
                   'template' => 'block/example.html.twig',
                    'label' => false,
                    'options' => [
                        'object' => $this->getSubject(),
                    ],
                ])
         ->end();
}
```

Template:
```twig
{#block/example.html.twig#}
<div class="col-md-3">
    <div class="info-box bg-green">
        <span class="info-box-icon"><i class="fa fa-thumbs-o-up"></i></span>

        <div class="info-box-content">
            <span class="info-box-text">Likes</span>
            <span class="info-box-number">41,410</span>

            <div class="progress">
                <div class="progress-bar" style="width: 70%"></div>
            </div>
            <span class="progress-description">
                    70% Increase in 30 Days
                  </span>
        </div>
    </div>
</div>
```

Result:

Bildschirmfoto 2020-02-01 um 11 17 20

Without overriding the template.

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I'm working on #5863

With the current implementation, the following code
image
give this result
image

This allow to easily add button, charts, data between some fields in the edit/create view. I added the `field_description`, `admin` and `object` parameters to allow people to use the same templates for Show and Form views.

Dunno which signature is better.
Something like 
```
->addTemplate('admin/sonata/button/api_button.html.twig', []);
```
could seems perfect, but
- Since it creates a form field, a unique form field name is needed every times. I could generate a uniq one every times, but since there is some reorders methods, the user should be able to reorder templates too. So it should be
```
->addTemplate('test', 'admin/sonata/button/api_button.html.twig', []);
```
- And then people could ask to be able to remove the form label, or to add an help text, ... or anything provided by the form, so either a fourth parameter is provided, either a signature
```
->addTemplate('test', [
    'template' => 'admin/sonata/button/api_button.html.twig',
    'parameters' => [],
])
```
is used, which I'm not happy with.

Any idea @sonata-project/contributors @jordisala1991 ?

Closes #5863.

## Changelog

```markdown
### Added
- `TemplateType` to display custom data in the edit/create view.
```

### CODE REVIEW

**Comment by jordisala1991** [comment]
Why do we need to treat it like a form field? Can't it be something different? Like we do the `with` to create form sections

**Comment by VincentLanglet** [comment]
> Why do we need to treat it like a form field? Can't it be something different? Like we do the `with` to create form sections

The two company I worked for was using this strategy to implement this feature, I feel like it was the more natural then.

Groups are rendered this way 
https://github.com/sonata-project/SonataAdminBundle/blob/78655c25923fa3ddf810d670d7b02fc447bb948f/src/Resources/views/CRUD/base_edit_form_macro.html.twig#L4-L43

- A foreach on each group
- Display the group title/box
- A foreach on each field inside the group
- Display the form_field_name with `{{ form_row(form[form_field_name]) }}`

What might be done instead is something like 
```
{% for form_field_name in form_group.fields %}
    {% if form[form_field_name] is defined %}
        {{ form_row(form[form_field_name]) }}
    {% else %}
        // Custom render
    {% end %}
{% else %} 
    {% if form_group.empty_message != false %} 
        {% if form_group.empty_message_translation_domain is defined and form_group.empty_message_translation_domain is same as(false) %} 
            <em>{{ form_group.empty_message }}</em> 
        {% else %} 
            <em>{{ form_group.empty_message|trans({}, form_group.empty_message_translation_domain ?? admin.translationDomain) }}</em> 
        {% endif %} 
    {% endif %} 
{% endfor %}
```

If you use this strategy, you won't benefit as easily from the classic "form design" with a label, an helper, etc and you won't benefit from custom form themes if you use it.

Also, if we don't work with a form type, we'll have some issue with the following logic in the FormMapper
```
    /**
     * @return static
     */
    public function reorder(array $keys): self
    {
        $this->getAdmin()->reorderFormGroup($this->getCurrentGroupName(), $keys);

        return $this;
    }

    public function get(string $key): FormBuilderInterface
    {
        $name = $this->sanitizeFieldName($key);

        return $this->formBuilder->get($name);
    }

    public function has(string $key): bool
    {
        $key = $this->sanitizeFieldName($key);

        return $this->formBuilder->has($key);
    }

    public function keys(): array
    {
        return array_keys($this->formBuilder->all());
    }

    /**
     * @return static
     */
    public function remove(string $key): self
    {
        $this->getAdmin()->removeFormFieldDescription($key);
        $this->getAdmin()->removeFieldFromFormGroup($key);

        $sanitizedKey = $this->sanitizeFieldName($key);
        $this->formBuilder->remove($sanitizedKey);

        return $this;
    }
```
Let's say I have a field `foo` and a template `templateBar`.
- has('templateBar') returns false and get('templateBar') throw an exception (and we cannot change this since `get` have to return a FormBuilderInterface)
- keys() returns ['foo']
- dunno if remove('templateBar') will work
- but still, if you want to reorder your fields and yours templates you'll have to call `reorder(['templateBar', 'foo'])`
I don't see an easy way to have some consistency in those methods if templates are not form fields...

**Comment by VincentLanglet** [comment]
So WDYT I should do @jordisala1991 ? Are you ok with the TemplateType given the previous message ?

**Comment by jordisala1991** [comment]
Yes, go ahead with the form type

**Review by core23 — DISMISSED** [review]

File: src/Form/Type/TemplateType.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace Sonata\AdminBundle\Form\Type;
+
+use Sonata\AdminBundle\FieldDescription\FieldDescriptionInterface;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class TemplateType extends AbstractType
```
Comment (core23) [inline src/Form/Type/TemplateType.php]:
```suggestion
final class TemplateType extends AbstractType
```

**Review by core23 — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Is it ok for u @jordisala1991 ?

**Review by jordisala1991 — APPROVED** [review]

**Comment by jordisala1991** [comment]
Thanks!

**Pull request merged**